File: src/main/java/com/codebutler/farebot/ui/NumberPickerPreference.java
/*
 * NumberPickerPreference.java
 * Copyright 2016 Michael Farrell <micolous+git@gmail.com>
 *
 * Based on EditTextPreference from Android Open Source Project:
 * Copyright (C) 2007 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.codebutler.farebot.ui;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.TypedArray;
import android.os.Parcel;
import android.os.Parcelable;
import android.preference.DialogPreference;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewParent;
import android.widget.NumberPicker;


/**
 * Implements a preference which allows the user to pick a number.
 */

public class NumberPickerPreference extends DialogPreference {
    private NumberPicker mNumberPicker;
    private int mValue = 0;
    private boolean mValueSet;
    private static final String NPP_SCHEMA = "http://micolous.github.io/metrodroid/schemas/number-picker-preference";

    public NumberPickerPreference(Context context, AttributeSet attributes, int defStyleAttr) {
        super(context, attributes, defStyleAttr);

        mNumberPicker = new NumberPicker(context, attributes);

        mNumberPicker.setId(android.R.id.edit);
        mNumberPicker.setEnabled(true);
        mNumberPicker.setMinValue(attributes.getAttributeIntValue(NPP_SCHEMA, "minValue", 0));
        mNumberPicker.setMaxValue(attributes.getAttributeIntValue(NPP_SCHEMA, "maxValue", 100));
    }

    public NumberPickerPreference(Context context, AttributeSet attributes) {
        this(context, attributes, android.R.attr.editTextPreferenceStyle);
    }

    public NumberPickerPreference(Context context) {
        this(context, null);
    }



    /**
     * Saves the text to the {@link SharedPreferences}.
     *
     * @param value The text to save
     */
    public void setValue(int value) {
        // Always persist/notify the first time.
        final boolean changed = mValue != value;
        if (changed || !mValueSet) {
            mValue = value;
            mValueSet = true;
            persistInt(value);
            if(changed) {
                notifyDependencyChange(shouldDisableDependents());
                notifyChanged();
            }
        }
    }

    /**
     * Gets the text from the {@link SharedPreferences}.
     *
     * @return The current preference value.
     */
    public int getValue() {
        return mValue;
    }


    @Override
    protected View onCreateDialogView() {
        mNumberPicker.setValue(getValue());
        ViewParent parent = mNumberPicker.getParent();
        if (parent != null) {
            ((ViewGroup)parent).removeView(mNumberPicker);
        }
        return mNumberPicker;
    }

    @Override
    protected void onDialogClosed(boolean positiveResult) {
        super.onDialogClosed(positiveResult);

        if (positiveResult) {
            int value = mNumberPicker.getValue();
            if (callChangeListener(value)) {
                setValue(value);
            }
        }
    }

    @Override
    protected Object onGetDefaultValue(TypedArray a, int index) {
        return a.getInteger(index, 0);
    }

    @Override
    protected void onSetInitialValue(boolean restoreValue, Object defaultValue) {
        setValue(restoreValue ? getPersistedInt(mValue) : ((Integer) defaultValue).intValue());
    }

    /**
     * Returns the {@link NumberPicker} widget that will be shown in the dialog.
     *
     * @return The {@link NumberPicker} widget that will be shown in the dialog.
     */
    public NumberPicker getNumberPicker() {
        return mNumberPicker;
    }

    @Override
    protected Parcelable onSaveInstanceState() {
        final Parcelable superState = super.onSaveInstanceState();
        if (isPersistent()) {
            // No need to save instance state since it's persistent
            return superState;
        }

        final SavedState myState = new SavedState(superState);
        myState.value = getValue();
        return myState;
    }

    @Override
    protected void onRestoreInstanceState(Parcelable state) {
        if (state == null || !state.getClass().equals(SavedState.class)) {
            // Didn't save state for us in onSaveInstanceState
            super.onRestoreInstanceState(state);
            return;
        }

        SavedState myState = (SavedState) state;
        super.onRestoreInstanceState(myState.getSuperState());
        setValue(myState.value);
    }

    private static class SavedState extends BaseSavedState {
        int value;

        public SavedState(Parcel source) {
            super(source);
            value = source.readInt();
        }
        @Override
        public void writeToParcel(Parcel dest, int flags) {
            super.writeToParcel(dest, flags);
            dest.writeInt(value);
        }
        public SavedState(Parcelable superState) {
            super(superState);
        }
        public static final Parcelable.Creator<SavedState> CREATOR =
                new Parcelable.Creator<SavedState>() {
                    public SavedState createFromParcel(Parcel in) {
                        return new SavedState(in);
                    }
                    public SavedState[] newArray(int size) {
                        return new SavedState[size];
                    }
                };
    }
}

File: src/main/java/au/id/micolous/metrodroid/ui/NumberPickerPreference.kt
/*
 * NumberPickerPreference.kt
 * Copyright 2016 Michael Farrell <micolous+git@gmail.com>
 *
 * Based on EditTextPreference from Android Open Source Project:
 * Copyright (C) 2007 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package au.id.micolous.metrodroid.ui

import android.content.Context
import android.content.SharedPreferences
import android.content.res.TypedArray
import android.os.Build
import android.os.Parcelable
import androidx.preference.DialogPreference
import android.util.AttributeSet
import androidx.annotation.RequiresApi
import kotlinx.parcelize.Parcelize
import au.id.micolous.farebot.R


/**
 * Implements a preference which allows the user to pick a number.
 */

// Used from XML
@Suppress("unused")
class NumberPickerPreference : DialogPreference {
    val minValue: Int
    val maxValue: Int

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes) {
        minValue = attrs.getAttributeIntValue(NPP_SCHEMA, "minValue", 0)
        maxValue = attrs.getAttributeIntValue(NPP_SCHEMA, "maxValue", 100)
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        minValue = attrs.getAttributeIntValue(NPP_SCHEMA, "minValue", 0)
        maxValue = attrs.getAttributeIntValue(NPP_SCHEMA, "maxValue", 100)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        minValue = attrs.getAttributeIntValue(NPP_SCHEMA, "minValue", 0)
        maxValue = attrs.getAttributeIntValue(NPP_SCHEMA, "maxValue", 100)
    }

    /**
     * The current preference value.
     *
     * The text is saved to the [SharedPreferences]
     */
    var value = 0
        set(value) {
            val changed = field != value
            // Always persist/notify the first time.
            if (changed || !mValueSet) {
                field = value
                mValueSet = true
                persistInt(value)
                if (changed) {
                    notifyDependencyChange(shouldDisableDependents())
                    notifyChanged()
                }
            }
        }
    private var mValueSet: Boolean = false

    override fun onGetDefaultValue(a: TypedArray, index: Int): Any {
        return a.getInteger(index, 0)
    }

    override fun onSetInitialValue(restoreValue: Boolean, defaultValue: Any?) {
        value = if (restoreValue) getPersistedInt(value) else defaultValue as? Int ?: 0
    }

    override fun onSaveInstanceState(): Parcelable {
        val superState = super.onSaveInstanceState()
        if (isPersistent) {
            // No need to save instance state since it's persistent
            return superState
        }

        return SavedState(superState, value)
    }

    override fun onRestoreInstanceState(state: Parcelable?) {
        if (state == null || state.javaClass != SavedState::class.java || state !is SavedState) {
            // Didn't save state for us in onSaveInstanceState
            super.onRestoreInstanceState(state)
            return
        }

        super.onRestoreInstanceState(state.superState)
        value = state.value
    }

    override fun getDialogLayoutResource(): Int = R.layout.pref_number_picker

    @Parcelize
    private class SavedState (
            internal val superState: Parcelable,
            internal val value: Int): Parcelable

    companion object {
        private const val NPP_SCHEMA = "http://micolous.github.io/metrodroid/schemas/number-picker-preference"
    }
}
